src/safearray.rs

```diff
@@ -21,35 +25,101 @@ extern "system" {
     pub fn SafeArrayDestroy(psa: *mut SAFEARRAY) -> HRESULT;
 }
 
-pub fn get_string_array(arr: *mut SAFEARRAY) -> Result<Vec<String>, Error> {
-    let mut p_data = NULL;
+#[derive(Debug)]
+struct SafeArrayAccessor {
+    arr: *mut SAFEARRAY,
+    p_data: *mut c_void,
+    lower_bound: i32,
+    upper_bound: i32,
+}
+
+/// An accessor to SafeArray, which:
+/// 1. locks the array allows reading the data.
+/// 2. and unlocks the array once dropped.
+///
+impl SafeArrayAccessor {
+    pub fn new(arr: *mut SAFEARRAY) -> Result<Self, Error> {
+        let mut p_data = NULL;
+        let mut lower_bound: i32 = 0;
+        let mut upper_bound: i32 = 0;
 
-    let mut lower_bound: i32 = 0;
-    let mut upper_bound: i32 = 0;
+        unsafe {
+            check_hres(SafeArrayGetLBound(arr, 1, &mut lower_bound as _))?;
+            check_hres(SafeArrayGetUBound(arr, 1, &mut upper_bound as _))?;
+            check_hres(SafeArrayAccessData(arr, &mut p_data))?;
+        }
 
-    unsafe {
-        check_hres(SafeArrayGetLBound(arr, 1, &mut lower_bound as _))?;
-        check_hres(SafeArrayGetUBound(arr, 1, &mut upper_bound as _))?;
-        check_hres(SafeArrayAccessData(arr, &mut p_data))?;
+        Ok(Self {
+            arr,
+            p_data,
+            lower_bound,
+            upper_bound,
+        })
     }
 
-    // We know that we expect an array of BSTRs.
-    let p_data: *mut BSTR = p_data as _;
+    /// Return a slice which can access the data of the array.
+    ///
+    /// # Safety
+    ///
+    /// This function is unsafe as it is the caller's responsibility to verify that the array is
+    /// of items of type T.
+    pub unsafe fn as_slice<T>(&self) -> &[T] {
+        let p_data: *mut T = self.p_data as _;
+
+        // upper_bound can be -1, in which case the array is empty and we will return a 0 length slice.
+        let data_slice = unsafe { slice::from_raw_parts(p_data, (self.upper_bound + 1) as usize) };
+
+        &data_slice[(self.lower_bound as usize)..]
+    }
+}
 
-    // upper_bound can be -1, in which case the array is empty and we will do nothing.
-    let data_slice = unsafe { slice::from_raw_parts(p_data, (upper_bound + 1) as usize) };
+impl Drop for SafeArrayAccessor {
+    fn drop(&mut self) {
+        // TOOD: Should we handle errors in some way?
+        unsafe {
+            let _result = check_hres(SafeArrayUnaccessData(self.arr));
+        }
+    }
+}
+
+pub fn safe_array_to_vec_of_strings(arr: *mut SAFEARRAY) -> Result<Vec<String>, Error> {
+    let items = safe_array_to_vec(arr, VT_BSTR)?;
+
+    let string_items = items
+        .into_iter()
+        .map(|item| match item {
+            Variant::String(s) => s,
+            _ => unreachable!(),
+        })
+        .collect();
+
+    Ok(string_items)
+}
+
+pub fn safe_array_to_vec(arr: *mut SAFEARRAY, item_type: u32) -> Result<Vec<Variant>, Error> {
```

Comment:
Would it be possible to extend std::Vec instead?

`Vec::from_safearray<T>(&SafeArray)`?

Reply:
Well it doesn't really seem like it would be worth it (I guess it just goes against my personal style)